firmware/applications/SignBuddy/tsc.c

```diff
@@ -92,57 +205,83 @@ static int read_value(uint32_t *value)
   }
 
   __HAL_TSC_CLEAR_FLAG(&s.tsc, (TSC_FLAG_EOA | TSC_FLAG_MCE));
-
-
-  if (HAL_TSC_GroupGetStatus(&s.tsc, TSC_GROUP3_IDX) == TSC_GROUP_COMPLETED) {
-    *value = HAL_TSC_GroupGetValue(&s.tsc, TSC_GROUP3_IDX);
-  }
   return RET_OK;
 }
 
 static inline void sample_data(void)
 {
-  uint32_t value;
+  memset(s.touch_values, 0, sizeof(s.touch_values));
 
-  if (read_value(&value) == RET_OK) {
-    //LOG_DEBUG("tsc value: %lu cal: %lu\r\n", value, s.calibration_value);
-    s.touch_value = convert_value_to_touch(value);
+  if (run_sampler() == RET_OK) {
+    get_channel_values(CHANNEL0_MIN_GROUP, CHANNEL0_MAX_GROUP, 0);
   }
   else {
-    s.touch_value = 0;
+    LOG_WARN("WARNING: TSC MAX CNT REACHED\r\n");
   }
 
-  io_config();
+  io_config(CHANNEL1);
+  HAL_TSC_IODischarge(&s.tsc, ENABLE);
+  rtos_delay_ms(1);
+
+  if (run_sampler() == RET_OK) {
+    get_channel_values(CHANNEL1_MIN_GROUP, CHANNEL1_MAX_GROUP, CHANNEL0_CNT);
+  }
+  else {
+    LOG_WARN("WARNING: TSC MAX CNT REACHED\r\n");
+  }
+
+
+  io_config(CHANNEL0);
   HAL_TSC_IODischarge(&s.tsc, ENABLE);
 }
 
+static void calibrate_channel_values(uint8_t min_group, uint8_t max_group, uint8_t offset)
+{
+  uint32_t value;
+  int index = offset;
+
+  for (int i = min_group; i <= max_group; i++) {
+    get_group_value(i, &value);
+    s.calibration_values[index] += value / CALIBRATION_SAMPLES;
+    index++;
+  }
+}
+
 static void calibrate(void)
 {
-  uint32_t calibration_value = 0;
+  memset(s.calibration_values, 0, sizeof(s.calibration_values));
 
   for (int i = 0; i < CALIBRATION_SAMPLES; i++) {
     // TODO find bug
     // if log is added here, lower values are read (closer to what is seen during sampling)
-    io_config();
+    io_config(CHANNEL0);
     HAL_TSC_IODischarge(&s.tsc, ENABLE);
-
     rtos_delay_ms(10);
 
-    uint32_t value;
+    if (run_sampler() == RET_OK) {
+      calibrate_channel_values(CHANNEL0_MIN_GROUP, CHANNEL0_MAX_GROUP, 0);
+    }
+    else {
+      LOG_WARN("WARNING: TSC MAX CNT REACHED\r\n");
+      // set to max value
+      memset(s.calibration_values, 0xFF, sizeof(s.calibration_values));
+      return;
+    }
+
+    io_config(CHANNEL1);
+    HAL_TSC_IODischarge(&s.tsc, ENABLE);
+    rtos_delay_ms(10);
 
-    if (read_value(&value) == RET_OK) {
-      calibration_value += value / CALIBRATION_SAMPLES;
+    if (run_sampler() == RET_OK) {
```

Comment:
The same functionality is called twice here, maybe pass the CHANNEL, CHANNEL, and MIN_GROUP, MAX_GROUP as arguments to calibrate()